flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java

```diff
@@ -525,56 +545,102 @@ public static Integer toInt(BinaryStringData str) {
         }
-        return null;
+        throw numberFormatExceptionFor(str, "SHORT", "Overflow");
     }
 
-    public static Byte toByte(BinaryStringData str) {
-        Integer intValue = toInt(str);
-        if (intValue != null) {
-            byte result = intValue.byteValue();
-            if (result == intValue) {
-                return result;
-            }
+    public static byte toByte(BinaryStringData str) throws TableException {
+        int intValue = toInt(str);
+        byte result = (byte) intValue;
+        if (result == intValue) {
+            return result;
         }
-        return null;
+        throw numberFormatExceptionFor(str, "BYTE", "Overflow");
     }
 
-    public static Double toDouble(BinaryStringData str) {
+    public static double toDouble(BinaryStringData str) throws TableException {
         try {
-            return Double.valueOf(str.toString());
+            return Double.parseDouble(str.toString());
         } catch (NumberFormatException e) {
-            return null;
+            throw new TableException("Cannot parse STRING to DOUBLE", e);
         }
     }
 
-    public static Float toFloat(BinaryStringData str) {
+    public static float toFloat(BinaryStringData str) throws TableException {
         try {
-            return Float.valueOf(str.toString());
+            return Float.parseFloat(str.toString());
         } catch (NumberFormatException e) {
-            return null;
+            throw new TableException("Cannot parse STRING to FLOAT", e);
+        }
+    }
+
+    private static TableException numberFormatExceptionFor(
+            StringData input, String targetType, @Nullable String reason) {
+        if (reason == null) {
+            return new TableException(
+                    "Cannot parse STRING to " + targetType,
+                    new NumberFormatException("For input string: \"" + input + "\""));
+        }
+
+        return new TableException(
+                "Cannot parse STRING to " + targetType,
+                new NumberFormatException("For input string: \"" + input + "\". " + reason));
+    }
+
+    public static int toDate(BinaryStringData str) throws TableException {
+        Integer date = DateTimeUtils.dateStringToUnixDate(str.toString());
+        if (date == null) {
+            throw new TableException("Cannot parse STRING to DATE");
         }
+
+        return date;
+    }
+
+    public static int toTime(BinaryStringData str) throws TableException {
+        Integer date = DateTimeUtils.timeStringToUnixDate(str.toString());
+        if (date == null) {
+            throw new TableException("Cannot parse STRING to TIME");
+        }
+
+        return date;
+    }
+
+    public static TimestampData toTimestamp(BinaryStringData str) throws TableException {
+        TimestampData timestamp = DateTimeUtils.toTimestampData(str.toString());
+        if (timestamp == null) {
+            throw new TableException("Cannot parse STRING to TIMESTAMP");
```

Comment:
Probably it makes sense to specify what string cannot be parsed like it is done for `toBoolean`.